daemon/cmd/app.go

```diff
@@ -0,0 +1,224 @@
+			return d.ipcache, d.mtuConfig
+		},
+	),
+
+	fx.Invoke(
+		preInit,
+		registerDaemonStart,
+		endpointManagerInit,
+	),
+)
+
+func runApp(cmd *cobra.Command) {
+	// Create a top-level context for the daemon. This is cancelled by the signal handler
+	// in cleanup.go and will trigger stop.
+	ctx, cancel := context.WithCancel(server.ServerCtx)
+
+	initEnv(cmd)
+
+	app := fx.New(
+		fx.StartTimeout(startTimeout),
+		fx.StopTimeout(stopTimeout),
+		fx.WithLogger(newAppLogger),
+
+		fx.Supply(
+			fx.Annotate(ctx, fx.As(new(context.Context))),
+			cancel),
+
+		fx.Supply(option.Config),
+
+		fx.Provide(
+			iptables.NewIptablesManager,
+			newEndpointManager,
+		),
+
+		// The order in which the modules are declared denotes the invoke and
+		// OnStart order.
+		linuxDatapathModule,
+		daemonModule,
+		wireguard.Module,
+		optional(option.Config.EnableIPMasqAgent, ipmasqAgentModule),
+
+		fx.Invoke(writeDotGraph),
+	)
+
+	if app.Err() != nil {
+		log.WithError(app.Err()).Fatal("Failed to initialize daemon")
+	}
+
+	if option.Config.DotGraphOutputFile != "" {
+		log.WithField("file", option.Config.DotGraphOutputFile).Infof("Wrote dot graph and now exiting without starting")
+		os.Exit(0)
+	}
+
+	startCtx, cancel := context.WithTimeout(context.Background(), app.StartTimeout())
+	defer cancel()
+
+	if err := app.Start(startCtx); err != nil {
+		if ctx.Err() != nil {
+			// Context was cancelled during startup, e.g. due to interrupt. Exit
+			// normally in this case.
+			os.Exit(0)
+		}
+		os.Exit(1)
+	}
+
+	// Wait until the daemon context is cancelled. This is done via the
+	// cleaner on receipt of an interrupt.
+	<-ctx.Done()
+
+	stopCtx, cancel := context.WithTimeout(context.Background(), app.StopTimeout())
+	defer cancel()
+
+	if err := app.Stop(stopCtx); err != nil {
+		log.WithError(err).Fatal("Failed to stop daemon")
+	}
+}
+
+func optional(flag bool, opts ...fx.Option) fx.Option {
+	if flag {
+		return fx.Options(opts...)
+	}
+	return fx.Invoke()
+}
+
+func preInit() error {
+	log.Info("Initializing daemon")
+	option.Config.RunMonitorAgent = true
+
+	if err := enableIPForwarding(); err != nil {
+		return fmt.Errorf("error when enabling sysctl parameters: %w", err)
+	}
+	if k8s.IsEnabled() {
+		bootstrapStats.k8sInit.Start()
+		if err := k8s.Init(option.Config); err != nil {
+			return fmt.Errorf("unable to initialize Kubernetes subsystem: %w", err)
+		}
+		bootstrapStats.k8sInit.End(true)
+	}
+	return nil
+}
+
+func registerDaemonStart(lc fx.Lifecycle, d *Daemon, restoredEndpoints *endpointRestoreState) {
+	lc.Append(fx.Hook{
+		OnStart: func(context.Context) error {
+			go runDaemon(d, restoredEndpoints)
+			return nil
+		},
+	})
+}
+
+func newEndpointManager(ctx context.Context) *endpointmanager.EndpointManager {
+	return WithDefaultEndpointManager(ctx, endpoint.CheckHealth)
+}
+
+func endpointManagerInit(lc fx.Lifecycle, em *endpointmanager.EndpointManager, ipcache *ipcache.IPCache, d *Daemon) {
```

Comment:
Yep you're identifying one of the sins I made here, which was to use the lifecycle from invokes rather than from constructors. I probably should've pointed it out explicitly as comments that what I'm doing here is hack due to not wanting to change existing code too much. Perhaps I should go a little bit further in this PR and make sure to move all lifecycle hook appends into constructors already in order to avoid introducing bad style from the get-go.

I think it's time I'll start writing some sort of "effective fx" guide for where to put things and what to avoid since the fx docs isn't very explicit about this.